In [3]:
%pip install pydantic-xml[lxml]

  Using cached pydantic_xml-2.14.3-py3-none-any.whl.metadata (5.1 kB)
Using cached pydantic_xml-2.14.3-py3-none-any.whl (40 kB)

[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [10]:
%pip install rich


[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


## Classes for XML Data

In [68]:
from typing import Dict
from pydantic import Field
from pydantic_xml import BaseXmlModel, attr, element


class EcuSystemFamily(BaseXmlModel):
    ref: str = attr("ref")
    name: str = ""

class EcuSystemExecution(BaseXmlModel):
    ref: str = attr("ref")
    name: str = ""

class ServerExecution(BaseXmlModel):
    ref: str = attr("ref")
    name: str = ""

class ImportedRangeRef(BaseXmlModel):
    ref: str = attr("ref")
    name: str = ""

class ImportedSimpleParameter(BaseXmlModel):
    name: str = element("Name")
    description: str = element("Description")
    Unit: str = element("Unit")
    rangeRef: ImportedRangeRef = element("ImportedRangeRef")
    UserFunction: str = element("UserFunction")


class PtImportedSimpleParameter(BaseXmlModel, tag="PtImportedSimpleParameter"):
    name: str = element("Name")
    ecuSystemFamily: EcuSystemFamily = element("EcuSystemFamily")
    ecuSystemExecution: EcuSystemExecution = element("EcuSystemExecution")
    serverExecution: ServerExecution = element("ServerExecution")
    importedSimpleParameter: ImportedSimpleParameter = element("ImportedSimpleParameter")

    class Config:
        xml_ns = {
            "xsi": "http://www.w3.org/2001/XMLSchema-instance"
        }

In [ ]:
from pathlib import Path
from lxml import etree
from rich import print

# Read as bytes to handle encoding declaration correctly
with Path("./data/Function-Parameters/APS2/Imported/Parameters/PtImportedSimpleParameter_ExtraDryingPerformance.xml").open("rb") as f:
    xml_data = f.read()


tree = etree.fromstring(xml_data)

# Now parse it
obj = PtImportedSimpleParameter.from_xml(xml_data)

# Access like normal
print(obj)


PtImportedSimpleParameter(
    name='ExtraDryingPerformance',
    ecuSystemFamily=EcuSystemFamily(ref='PtEcuSystemFamily', name='APS'),
    ecuSystemExecution=EcuSystemExecution(ref='PtEcuSystemExecution', name='APS2'),
    serverExecution=ServerExecution(ref='PtServerExecution', name='APS2'),
    importedSimpleParameter=ImportedSimpleParameter(
        name='ExtraDryingPerformance',
        description='Parameter to select different drying performance for high capacity device\r\n1 -> old drying 
performance setting\r\n2 -> new drying performance setting\r\n3 to 4 -> reserved for Future, new drying performance
setting is used',
        Unit='-',
        rangeRef=ImportedRangeRef(ref='PtImportedRange', name='ExtraDryingPerformanceRange'),
        UserFunction='193'
    )
)

In [78]:
print(obj.to_xml())

b'<PtImportedSimpleParameter><Name>ExtraDryingPerformance</Name><EcuSystemFamily 
ref="PtEcuSystemFamily">APS</EcuSystemFamily><EcuSystemExecution 
ref="PtEcuSystemExecution">APS2</EcuSystemExecution><ServerExecution 
ref="PtServerExecution">APS2</ServerExecution><ImportedSimpleParameter><Name>ExtraDryingPerformance</Name><Descript
ion>Parameter to select different drying performance for high capacity device&#13;\n1 -&gt; old drying performance 
setting&#13;\n2 -&gt; new drying performance setting&#13;\n3 to 4 -&gt; reserved for Future, new drying performance
setting is used</Description><Unit>-</Unit><ImportedRangeRef 
ref="PtImportedRange">ExtraDryingPerformanceRange</ImportedRangeRef><UserFunction>193</UserFunction></ImportedSimpl
eParameter></PtImportedSimpleParameter>'

## Ranges

In [79]:
from typing import List
from pydantic import Field
from pydantic_xml import BaseXmlModel, attr, element


class EcuSystemFamily(BaseXmlModel):
    ref: str = attr("ref")
    name: str = ""

class EcuSystemExecution(BaseXmlModel):
    ref: str = attr("ref")
    name: str = ""

class ServerExecution(BaseXmlModel):
    ref: str = attr("ref")
    name: str = ""

class ImportedDiscreteValue(BaseXmlModel):
    name: str = element("Name")


class PtImportedRange(BaseXmlModel, tag="PtImportedRange"):
    name: str = element("Name")
    ecuSystemFamily: EcuSystemFamily = element("EcuSystemFamily")
    ecuSystemExecution: EcuSystemExecution = element("EcuSystemExecution")
    serverExecution: ServerExecution = element("ServerExecution")
    importedDiscreteValue: List[ImportedDiscreteValue] = element("ImportedDiscreteValue")

    class Config:
        xml_ns = {
            "xsi": "http://www.w3.org/2001/XMLSchema-instance"
        }

In [82]:
from pathlib import Path
from lxml import etree
from rich import print

# Read as bytes to handle encoding declaration correctly
with Path("./data/Function-Parameters/APS2/Imported/Ranges/PtImportedRange_ExtraDryingPerformanceRange.xml").open("rb") as f:
    xml_data = f.read()

print(xml_data)
tree = etree.fromstring(xml_data)

# Now parse it
obj = PtImportedRange.from_xml(xml_data)

# Access like normal
print(obj)


b'<?xml version="1.0" encoding="UTF-8" standalone="no"?>\r\n<PtImportedRange 
xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
xsi:noNamespaceSchemaLocation="../../../../../../../../../Dev/schema/sds/ecuSystem/sdp/propertyRange/PtImportedRang
e.xsd">\r\n    <Name>ExtraDryingPerformanceRange</Name>\r\n    <EcuSystemFamily 
ref="PtEcuSystemFamily">APS</EcuSystemFamily>\r\n    <EcuSystemExecution 
ref="PtEcuSystemExecution">APS2</EcuSystemExecution>\r\n    <ServerExecution 
ref="PtServerExecution">APS2</ServerExecution>\r\n    <ImportedDiscreteValue>\r\n        <Name>Off</Name>\r\n    
</ImportedDiscreteValue>\r\n    <ImportedDiscreteValue>\r\n        <Name>Max</Name>\r\n    
</ImportedDiscreteValue>\r\n</PtImportedRange>\r\n'

PtImportedRange(
    name='ExtraDryingPerformanceRange',
    ecuSystemFamily=EcuSystemFamily(ref='PtEcuSystemFamily', name='APS'),
    ecuSystemExecution=EcuSystemExecution(ref='PtEcuSystemExecution', name='APS2'),
    serverExecution=ServerExecution(ref='PtServerExecution', name='APS2'),
    importedDiscreteValue=[ImportedDiscreteValue(name='Off'), ImportedDiscreteValue(name='Max')]
)

## Property Group

In [18]:
from typing import List, Optional
from pydantic_xml import BaseXmlModel, attr, element


class NamePresentation(BaseXmlModel):
    edt: Optional[str] = attr("edt")
    value: str = ""


class Description(BaseXmlModel):
    edt: Optional[str] = attr("edt")
    value: str = ""


class ProductVariantConditionRef(BaseXmlModel):
    ref: Optional[str] = attr("ref")
    value: str = ""


class Server(BaseXmlModel):
    canAddress: str = element("CanAddress")
    propertyName: str = element("PropertyName")


class Property(BaseXmlModel):
    server: Server = element("Server")


class PropertyGroup(BaseXmlModel):
    name: str = element("Name")
    namePresentation: NamePresentation = element("NamePresentation")
    description: Optional[Description] = element("Description", default=None)  # ✅ Added
    property: List[Property] = element("Property")


class FunctionPropertyGroup(BaseXmlModel, tag="FunctionPropertyGroup"):
    name: str = element("Name")
    namePresentation: NamePresentation = element("NamePresentation")
    description: Optional[Description] = element("Description", default=None)
    productVariantConditionRef: Optional[ProductVariantConditionRef] = element("ProductVariantConditionRef", default=None)
    ufNumber: List[int] = element("UFNumber")

    propertyGroup: Optional[List[PropertyGroup]] = element("PropertyGroup", default=None)
    property: Optional[List[Property]] = element("Property", default=None)

    class Config:
        xml_ns = {
            "xsi": "http://www.w3.org/2001/XMLSchema-instance"
        }

In [34]:
from pathlib import Path
from lxml import etree
from rich import print

# Read as bytes to handle encoding declaration correctly
with Path("./data/Function-Parameters/FunctionViewAdjust/FunctionPropertyGroup_EngineSpeedControl.xml").open("rb") as f:
    xml_data = f.read()

tree = etree.fromstring(xml_data)

# Now parse it
obj = FunctionPropertyGroup.from_xml(xml_data)

# Access like normal
print(obj)


FunctionPropertyGroup(
    name='EngineSpeedControl',
    namePresentation=NamePresentation(edt='nfTxt', value='Engine speed control'),
    description=Description(
        edt='fTxt',
        value='Here you can adjust parameters that affect engine speed control. The functions that is affected is, 
Hand throttle, Limited hand throttle (mode 1), Raised idling speed (mode 2), Predefined (locked) engine speed (mode
3) and Remote engine speed control (mode 4).'
    ),
    productVariantConditionRef=ProductVariantConditionRef(
        ref='ProductVariantCondition',
        value='Without_BWE_Without_ElectricMotor'
    ),
    ufNumber=[93, 221, 222, 275],
    propertyGroup=None,
    property=[Property(server=Server(canAddress='00', propertyName='PTOGearDeac'))]
)

## Physical Quantity

In [2]:
from typing import List, Optional
from pydantic_xml import BaseXmlModel, attr, element


class NamePresentation(BaseXmlModel):
    edt: Optional[str] = attr("edt")
    value: str = ""


class Unit(BaseXmlModel):
    name: str = element("Name")
    namePresentation: NamePresentation = element("NamePresentation")
    factor: Optional[float] = element("Factor", default=None)


class StandardUnit(BaseXmlModel):
    ref: str = attr("ref")
    value: str = ""


class PhysicalQuantity(BaseXmlModel, tag="PhysicalQuantity"):
    name: str = element("Name")
    namePresentation: NamePresentation = element("NamePresentation")
    standardUnit: StandardUnit = element("StandardUnit")
    unit: List[Unit] = element("Unit")

    class Config:
        xml_ns = {
            "xsi": "http://www.w3.org/2001/XMLSchema-instance"
        }

In [ ]:
from pathlib import Path
from lxml import etree
from rich import print

# Read as bytes to handle encoding declaration correctly
with Path("./data/Function-Parameters/PhysicalQuantity/PhysicalQuantity_Liters_per_minute.xml").open("rb") as f:
    xml_data = f.read()

print(xml_data)

# Now parse it
obj = PhysicalQuantity.from_xml(xml_data)

# Access like normal
print(obj)


b'<?xml version="1.0" encoding="UTF-8" standalone="no"?>\r\n<PhysicalQuantity 
xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
xsi:noNamespaceSchemaLocation="../../../Dev/schema/PhysicalQuantity/PhysicalQuantity.xsd">\r\n\t<Name>Liters_per_mi
nute</Name>\r\n\t<NamePresentation edt="nfTxt">l/min</NamePresentation>\r\n  <StandardUnit 
ref="Unit">l/min</StandardUnit>\r\n  <Unit>\r\n    <Name>l/min</Name>\r\n    <NamePresentation 
edt="nfTxt">l/min</NamePresentation>\r\n  </Unit>\r\n</PhysicalQuantity>\r\n'

PhysicalQuantity(
    name='Liters_per_minute',
    namePresentation=NamePresentation(edt='nfTxt', value='l/min'),
    standardUnit=StandardUnit(ref='Unit', value='l/min'),
    unit=[Unit(name='l/min', namePresentation=NamePresentation(edt='nfTxt', value='l/min'), factor=None)]
)

In [26]:
from typing import List, Optional
from pydantic_xml import BaseXmlModel, attr, element


class NamePresentation(BaseXmlModel):
    edt: Optional[str] = attr("edt")
    value: str = ""


class FunctionReference(BaseXmlModel):
    ref: str = attr("ref")
    value: str = ""


class Content(BaseXmlModel):
    functionPropertyGroup: Optional[List[FunctionReference]] = element("FunctionPropertyGroup", default=None)
    functionGuidedMethodControlGroup: Optional[List[FunctionReference]] = element("FunctionGuidedMethodControlGroup", default=None)
    functionGuidedMethodCalibrateGroup: Optional[List[FunctionReference]] = element("FunctionGuidedMethodCalibrateGroup", default=None)


class Group(BaseXmlModel):
    name: str = element("Name")
    namePresentation: NamePresentation = element("NamePresentation")
    group: Optional[List["Group"]] = element("Group", default=None)  # recursive
    content: Optional[Content] = element("Content", default=None)


class FunctionView(BaseXmlModel, tag="FunctionView"):
    name: str = element("Name")
    namePresentation: NamePresentation = element("NamePresentation")
    group: List[Group] = element("Group")

    class Config:
        xml_ns = {
            "xsi": "http://www.w3.org/2001/XMLSchema-instance"
        }



In [111]:
from pathlib import Path
from lxml import etree
from rich import print

# Read as bytes to handle encoding declaration correctly
with Path("./data/Function-Parameters/FunctionView_FunctionAdjustTree.xml").open("rb") as f:
    xml_data = f.read()

tree = etree.fromstring(xml_data)

# Now parse it
obj = FunctionView.from_xml(xml_data)

# Access like normal
print(obj)


FunctionView(
    name='FunctionAdjustTree',
    namePresentation=NamePresentation(edt='nfTxt', value='Present the vehicle identification number here'),
    group=[
        Group(
            name='BodyworkAdaptation',
            namePresentation=NamePresentation(edt='nfTxt', value='Bodywork adaptation'),
            group=[
                Group(
                    name='Preparation',
                    namePresentation=NamePresentation(edt='nfTxt', value='Preparation'),
                    group=None,
                    content=Content(
                        functionPropertyGroup=[
                            FunctionReference(ref='FunctionPropertyGroup', value='ExternalCANInterface')
                        ],
                        functionGuidedMethodControlGroup=None,
                        functionGuidedMethodCalibrateGroup=None
                    )
                )
            ],
            content=Content(
                functionPropertyGroup=None,
                functionGuidedMethodControlGroup=[
                    FunctionReference(ref='FunctionGuidedMethodControlGroup', value='FootPlate'),
                    FunctionReference(
                        ref='FunctionGuidedMethodControlGroup',
                        value='BodyworkInterfaceConfiguration'
                    ),
                    FunctionReference(
                        ref='FunctionGuidedMethodControlGroup',
                        value='DisplayInstrumentSignalsAndSimulatedTachographSignals'
                    ),
                    FunctionReference(
                        ref='FunctionGuidedMethodControlGroup',
                        value='HandleSwitchAndSwitchModules'
                    ),
                    FunctionReference(ref='FunctionGuidedMethodControlGroup', value='OptionalOutputSignals')
                ],
                functionGuidedMethodCalibrateGroup=None
            )
        ),
        Group(
            name='Chassis',
            namePresentation=NamePresentation(edt='nfTxt', value='Chassis'),
            group=[
                Group(
                    name='PressurizedAirManagement',
                    namePresentation=NamePresentation(edt='nfTxt', value='Pressurized air management'),
                    group=None,
                    content=Content(
                        functionPropertyGroup=None,
                        functionGuidedMethodControlGroup=[
                            FunctionReference(
                                ref='FunctionGuidedMethodControlGroup',
                                value='PressurizedAirManagement'
                            ),
                            FunctionReference(ref='FunctionGuidedMethodControlGroup', value='AirCompression')
                        ],
                        functionGuidedMethodCalibrateGroup=None
                    )
                ),
                Group(
                    name='Articulationsystem',
                    namePresentation=NamePresentation(edt='nfTxt', value='Articulation system'),
                    group=None,
                    content=Content(
                        functionPropertyGroup=[
                            FunctionReference(ref='FunctionPropertyGroup', value='Articulationcontrolsystem')
                        ],
                        functionGuidedMethodControlGroup=None,
                        functionGuidedMethodCalibrateGroup=None
                    )
                ),
                Group(
                    name='Brake',
                    namePresentation=NamePresentation(edt='nfTxt', value='Brake'),
                    group=None,
                    content=Content(
                        functionPropertyGroup=None,
                        functionGuidedMethodControlGroup=[
                            FunctionReference(
                                ref='FunctionGuidedMethodControlGroup',
                                value='BodyRequestedBraking'
                        

## OUTPUT FORMATS

In [ ]:
from typing import List
from pydantic_xml import BaseXmlModel, attr, element


class RefElement(BaseXmlModel):
    ref: str = attr("ref")
    value: str = ""


class PtSDP3Range(BaseXmlModel, tag="PtSDP3Range"):
    name: str = element("Name")
    ecuSystemFamily: RefElement = element("EcuSystemFamily")
    ecuSystemExecution: RefElement = element("EcuSystemExecution")
    serverExecution: RefElement = element("ServerExecution")
    importedRangeRef: RefElement = element("ImportedRangeRef")
    discreteValueRef: List[RefElement] = element("DiscreteValueRef")

    class Config:
        xml_ns = {
            "xsi": "http://www.w3.org/2001/XMLSchema-instance"
        }

In [ ]:
from pydantic_xml import BaseXmlModel, attr, element
from typing import Optional


class RefElement(BaseXmlModel):
    ref: str = attr("ref")
    value: str = ""


class PresentationText(BaseXmlModel):
    edt: Optional[str] = attr("edt")
    value: str = ""


class PtSDP3Parameter(BaseXmlModel, tag="PtSDP3Parameter"):
    name: str = element("Name")
    ecuSystemFamily: RefElement = element("EcuSystemFamily")
    ecuSystemExecution: RefElement = element("EcuSystemExecution")
    serverExecution: RefElement = element("ServerExecution")
    categoryCondition: RefElement = element("CategoryCondition")
    namePresentation: PresentationText = element("NamePresentation")
    description: PresentationText = element("Description")
    physicalQuantity: RefElement = element("PhysicalQuantity")
    importedSimpleParameterRef: RefElement = element("ImportedSimpleParameterRef")

    class Config:
        xml_ns = {
            "xsi": "http://www.w3.org/2001/XMLSchema-instance"
        }

In [8]:
%pip install tiktoken

  Using cached tiktoken-0.9.0-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (6.7 kB)
  Using cached regex-2024.11.6-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (40 kB)
Using cached tiktoken-0.9.0-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.2 MB)
Using cached regex-2024.11.6-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (796 kB)

[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [10]:
%pip install transformers

  Using cached filelock-3.18.0-py3-none-any.whl.metadata (2.9 kB)
  Using cached huggingface_hub-0.29.3-py3-none-any.whl.metadata (13 kB)
  Using cached tokenizers-0.21.1-cp39-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (6.8 kB)
  Using cached safetensors-0.5.3-cp38-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.8 kB)
  Using cached fsspec-2025.3.0-py3-none-any.whl.metadata (11 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.2/10.2 MB 184.9 MB/s eta 0:00:00
Using cached huggingface_hub-0.29.3-py3-none-any.whl (468 kB)
Using cached safetensors-0.5.3-cp38-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (471 kB)
Using cached tokenizers-0.21.1-cp39-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.0 MB)
Using cached filelock-3.18.0-py3-none-any.whl (16 kB)
Using cached fsspec-2025.3.0-py3-none-any.whl (193 kB)

[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to r

In [4]:
from transformers import AutoTokenizer

# Load the Qwen tokenizer
tokenizer = AutoTokenizer.from_pretrained("NovaSky-AI/Sky-T1-32B-Flash", trust_remote_code=True)

# Your prompt text
prompt = "Once upon a time in a faraway kingdom... and something something."

# Tokenize and count
input_ids = tokenizer(prompt)["input_ids"]
print(f"Token count: {len(input_ids)}")

/home/rajesh/Method_Engineer_Support_Tool/venv/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
None of PyTorch, TensorFlow >= 2.0, or Flax have been found. Models won't be available and only tokenizers, configuration and file/data utilities can be used.


Token count: 14


In [11]:
# load all the physical quantities from the ./data/Function-Parameters/PhysicalQuantity/ folder
from pathlib import Path

physical_quantity_files = Path("./data/Function-Parameters/PhysicalQuantity").glob(
    "*.xml"
)

# make sure their names start with PhysicalQuantity
physical_quantity_files = [
    file for file in physical_quantity_files if file.name.startswith("PhysicalQuantity")
]

physical_quantities = []
token_count = 0
print("total files: ", len(physical_quantity_files))

physical_quantities_details = ""

for file in physical_quantity_files:
    with file.open("rb") as f:
        xml_data = f.read()
        obj = PhysicalQuantity.from_xml(xml_data)

        json_data = obj.model_dump_json(indent=2)
        input_ids = tokenizer(json_data)["input_ids"]
        # add the name of the physical quantity to the details
        physical_quantities_details += "\n\nPHYSICAL QUANTITY " + obj.name + "\n"

        ## loop through the units and add them to the details
        for unit in obj.unit:
            physical_quantities_details += unit.name + "\n"

        token_count += len(input_ids)
        physical_quantities.append(obj)


# tokens for all physical quantities
tokens_count_physical_quantities = len(
    tokenizer(physical_quantities_details)["input_ids"]
)

print(f"Token count for all physical quantities: {tokens_count_physical_quantities}")

print(physical_quantities_details)



print(f"Token count: {token_count}")
# print(physical_quantities)

total files:  83
Token count for all physical quantities: 1250


PHYSICAL QUANTITY Rms
rpm/mg/str


PHYSICAL QUANTITY PartsPerMillion
ppm
-


PHYSICAL QUANTITY SNR
dB


PHYSICAL QUANTITY Volume_litres
l


PHYSICAL QUANTITY SignalStrength
%
-


PHYSICAL QUANTITY Time_year
-


PHYSICAL QUANTITY AngularVelocity
degree/s
rad/s


PHYSICAL QUANTITY Angle
degree
deg
rad


PHYSICAL QUANTITY Torque
Nm


PHYSICAL QUANTITY Flow_liters_per_seconds
l/s
ls


PHYSICAL QUANTITY Current
A
mA


PHYSICAL QUANTITY millimeter_per_degree
mm/deg


PHYSICAL QUANTITY Time_min
min
s


PHYSICAL QUANTITY Acceleration
m/s2
km/h/s


PHYSICAL QUANTITY Flow
l/h
l/min


PHYSICAL QUANTITY Impulses
I/km
I/m
pulses/m
pulses/km


PHYSICAL QUANTITY NumberOfRevolutionsPerSecond
rpm/s


PHYSICAL QUANTITY Energy_Wh
Wh


PHYSICAL QUANTITY Length_m
m


PHYSICAL QUANTITY Pressure
bar
Bar
mbar
mBar
Pa
kPa
kpa
hPa


PHYSICAL QUANTITY Length_cm
cm


PHYSICAL QUANTITY Length_mm
mm
m


PHYSICAL QUANTITY Volume_cm3
cm3


PHYSICAL QUAN

In [39]:
from openai import OpenAI

# Set OpenAI's API key and API base to use vLLM's API server.
openai_api_key = "EMPTY"
openai_api_base = "http://localhost:8000/v1"

client = OpenAI(
    api_key=openai_api_key,
    base_url=openai_api_base,
)

term = "DHSC_MinOffset"
description = "Function Parameter for RTDB_DHSC_MIN_OFFSET_E"
unit = "km/h"
descrete_values = [3, 4, 5, 7, 9, 11, 12, 15]

chat_response = client.chat.completions.create(
    model="NovaSky-AI/Sky-T1-32B-Flash",
    messages=[
        {
            "role": "system",
            "content": f"""
         You are a technical writer for an automotive company. You are familiar with various terms used in automotive documentation.
         Given a term and it's description, unit, you need to decide a type of physical quantity and the unit of measurement that is most appropriate for the term.

         ========================================
         ALL PHYSICAL QUANTITIES:
         {physical_quantities_details}

         ========================================
         RULES:

        1. Decide a type of physical quantity and the unit of measurement that is most appropriate for the term.
        2. Always choose the most specific physical quantity and unit of measurement.

        """,
        },
        {
            "role": "user",
            "content": f"""Term: {term}?
                Description: {description}
                descrete_values: {descrete_values}
                unit: {
                    unit
                }
                """,
        },
    ],
)
print("Chat response:", chat_response.choices[0].message.content)

Chat response: Given the term "DHSC_MinOffset?" with the description "Function Parameter for 
RTDB_DHSC_MIN_OFFSET_E" and discrete values [3, 4, 5, 7, 9, 11, 12, 15] in units of km/h, we need to determine the 
most appropriate physical quantity and unit of measurement.

### Analysis:
- **Term**: DHSC_MinOffset?
- **Description**: Function Parameter for RTDB_DHSC_MIN_OFFSET_E
- **Discrete Values**: [3, 4, 5, 7, 9, 11, 12, 15]
- **Unit**: km/h

### Decision:
The unit provided is "km/h," which is a unit of speed. Looking at the list of physical quantities, "Speed" is 
listed with units like "km/h," "m/s," "mph," etc. 

Given that the values are discrete and in "km/h," the most specific physical quantity that matches this description
is "Speed."

### Conclusion:
- **Physical Quantity**: Speed
- **Unit of Measurement**: km/h

This seems appropriate because "km/h" is a standard unit for measuring speed, and the values provided are 
consistent with possible speed settings or thresholds in a vehicle's system. 

Thus, the final answer is:

```
Physical Quantity: Speed
Unit: km/h
```

In [21]:
# load all the physical quantities from the ./data/Function-Parameters/PhysicalQuantity/ folder
from pathlib import Path

physical_quantity_files = Path("./data/Function-Parameters/FunctionViewAdjust").glob(
    "*.xml"
)

# make sure their names start with PhysicalQuantity
physical_quantity_files = [
    file
    for file in physical_quantity_files
    if file.name.startswith("FunctionPropertyGroup")
]

physical_quantities = []
token_count = 0
print("total files: ", len(physical_quantity_files))
funciton_properties_details = ""

for file in physical_quantity_files:
    with file.open("rb") as f:
        xml_data = f.read()
        obj = FunctionPropertyGroup.from_xml(xml_data)
        physical_quantities.append(obj)

        model_json = obj.model_dump_json(indent=2)

        input_ids = tokenizer(model_json)["input_ids"]
        token_count += len(input_ids)

        funciton_properties_details += "\n\nFUNCTION PROPERTY GROUP " + obj.name + "\n"
        if obj.property:
            for prop in obj.property:
                funciton_properties_details += prop.server.propertyName + "\n"
        if obj.propertyGroup:
            for prop_group in obj.propertyGroup:
                funciton_properties_details += prop_group.name + "\n"
                for prop in prop_group.property:
                    funciton_properties_details += prop.server.propertyName + "\n"

print(f"Token count: {token_count}")

# tokens for all physical quantities
tokens_count_function_properties = len(
    tokenizer(funciton_properties_details)["input_ids"]
)

print(f"Token count for all function properties: {tokens_count_function_properties}")
# print(physical_quantities)

total files:  164
Token count: 82831
Token count for all function properties: 15219


In [41]:
from openai import OpenAI

# Set OpenAI's API key and API base to use vLLM's API server.
openai_api_key = "EMPTY"
openai_api_base = "http://localhost:8000/v1"

client = OpenAI(
    api_key=openai_api_key,
    base_url=openai_api_base,
)

term = "DHSC_MinOffset"
description = (
    "Function Parameter for RTDB_DHSC_MIN_OFFSET_E"
)
descrete_values = []

chat_response = client.chat.completions.create(
    model="NovaSky-AI/Sky-T1-32B-Flash",
    messages=[
        {
            "role": "system",
            "content": f"""
         You are a technical writer for an automotive company. You are familiar with various terms used in automotive documentation.
         Given a term and it's description, you need to decide a group that we should use for the term.

         ========================================
         ALL Existing groups:
         {funciton_properties_details}

         ========================================
         RULES:

        1. Decide a group that we should use for the term.
        2. Always choose the most specific group.
        """,
        },
        {
            "role": "user",
            "content": f"""Term: {term}?
                Description: {description}
                descrete_values: {descrete_values}
                """,
        },
    ],
)
print("Chat response:", chat_response.choices[0].message.content)
print("prompt tokens:", chat_response.usage.prompt_tokens)
print("completion tokens:", chat_response.usage.completion_tokens)

Chat response: Given the term "DHSC_MinOffset" and its description as "Function Parameter for 
RTDB_DHSC_MIN_OFFSET_E," along with no discrete values provided, I need to determine the most specific group for 
this term from the existing groups listed.

Looking through the existing groups, I find a group specifically related to Downhill Speed Control:

```
FUNCTION PROPERTY GROUP DownhillSpeedControl
DHSC_Variant
DHSC_DefaultOffset
DHSCOffset
DHSC_MinOffset
DHSC
DHSCMinOffset
DhscBrakeLight
DhscBrakeLight
DHSCType
DHSC_AutoEngageType
```

Within this group, there are properties like `DHSC_MinOffset` which matches the term in question. Therefore, the 
most specific group for "DHSC_MinOffset" is the "DownhillSpeedControl" group.

**Decision:** Assign "DHSC_MinOffset" to the **DownhillSpeedControl** group.

prompt tokens: 15348

completion tokens: 197

### Function Adjust Tree file

In [34]:
from pathlib import Path
from lxml import etree
from rich import print

# Read as bytes to handle encoding declaration correctly
with Path("./data/Function-Parameters/FunctionView_FunctionAdjustTree.xml").open("rb") as f:
    xml_data = f.read()

tree = etree.fromstring(xml_data)

function_adjust_tree_info = ""

# Now parse it
obj = FunctionView.from_xml(xml_data)


model_json = obj.model_dump_json(indent=2)


input_ids = tokenizer(model_json)["input_ids"]

# convert the object to the info variable
for group in obj.group:
    function_adjust_tree_info += group.name + "\n"
    if group.group:
        for sub_group in group.group:
            function_adjust_tree_info += "-"+ sub_group.name + "\n"
            if sub_group.group:
                for sub_sub_group in sub_group.group:
                    function_adjust_tree_info += "--"+ sub_sub_group.name + "\n"

info_tokens = tokenizer(function_adjust_tree_info)["input_ids"]

print(f"Token count: {len(input_ids)}")
print(f"Token count: {len(info_tokens)}")

print(function_adjust_tree_info)



Token count: 6589

Token count: 193

BodyworkAdaptation
-Preparation
Chassis
-PressurizedAirManagement
-Articulationsystem
-Brake
-LevelControl
-FrontAxleSteering
-TagAxleSteering
-TirePressureMonitoring
-Trailer
Comfort
-DriverComfort
--ClimateSystem
--AudioAndNavigation
--CruiseControl
--Platooning
-PassengerComfort
--ClimateSystem
--PassengerInteraction
--Lighting
DriverAndVehicleFollowUp
Driver Assistance
DriverInformation
Maintenance
Performance
-PerformanceMode
-VehicleSpeed
PowerAndFluidSupply
Powertrain
-AllWheelDrive
-Clutch
-Engine
--EngineSpeedControl
-ExhaustTreatment
-Gearbox
-Hybrid
-PowerTakeOff
-AuxiliaryBrake
Safety
TheftProtection
Visibility
-Lighting
-Indication
-VisibilityCleaning
-Mirrors

In [42]:
from openai import OpenAI

# Set OpenAI's API key and API base to use vLLM's API server.
openai_api_key = "EMPTY"
openai_api_base = "http://localhost:8000/v1"

client = OpenAI(
    api_key=openai_api_key,
    base_url=openai_api_base,
)

term = "CameraSystems"

chat_response = client.chat.completions.create(
    model="NovaSky-AI/Sky-T1-32B-Flash",
    messages=[
        {
            "role": "system",
            "content": f"""
         You are a technical writer for an automotive company. You are familiar with various terms used in automotive documentation.
         Given a new group name, you need to decide the path where we should add this group in the Function Adjust Tree.

         ========================================
         Current Function Adjust Tree:
         {function_adjust_tree_info}

         ========================================
         RULES:

        1. Decide the path where we should add this group in the Function Adjust Tree.
        2. Always choose the most specific path.
        """,
        },
        {
            "role": "user",
            "content": f"""Term: {term}?
                """,
        },
    ],
)
print("Chat response:", chat_response.choices[0].message.content)
print("prompt tokens:", chat_response.usage.prompt_tokens)
print("completion tokens:", chat_response.usage.completion_tokens)

Chat response: Given the term "CameraSystems," we need to determine the most appropriate path in the Function 
Adjust Tree where this group should be added.

First, let's consider what "CameraSystems" entails. Camera systems in vehicles are typically used for various 
purposes such as:

- **Safety features**: Lane departure warning, blind-spot detection, backup cameras, etc.
- **Driver assistance**: Adaptive cruise control, parking assistance.
- **Visibility**: Helping drivers see better in various conditions.
- **Security**: Monitoring the vehicle and surroundings for theft protection.

Looking at the current Function Adjust Tree structure:

- **Safety**: This includes features that ensure the safety of the vehicle and its occupants. Since many camera 
systems are used for safety features, this could be a potential fit.
- **Driver Assistance**: This category includes systems that assist the driver in controlling the vehicle, such as 
adaptive cruise control and parking assistance, which often rely on camera systems.
- **Visibility**: This includes features related to improving the driver's view, such as backup cameras or cameras 
for monitoring blind spots.
- **Theft Protection**: Some camera systems might be used for security purposes, but this seems less common 
compared to their use in safety and driver assistance.

Given these considerations, the most specific and relevant path for "CameraSystems" would likely be under 
**Safety** or **Driver Assistance**, depending on the primary function of the camera systems in question.

However, since camera systems can serve multiple purposes, we should aim for the most inclusive yet specific 
category. In this case, **Safety** seems to be the broader category that encompasses many of the uses of camera 
systems in vehicles.

Therefore, the recommended path is:

```
Safety/CameraSystems
``` 

This placement ensures that camera systems are grouped with other safety-related features, making it easier to 
manage and understand their functions within the vehicle's system architecture.

prompt tokens: 307

completion tokens: 374

## Notes

1. With Physical Quantity(83 files), the raw xml data is around 15503 tokens
2. With FunctionGroup(164 files), the raw xml data is around (112478 tokens)
3. With FunctionAdjustTree(1 file), the raw xml data is around (6589 tokens)

## Reading Material

https://arxiv.org/pdf/2407.11963

https://qdrant.tech/articles/sparse-vectors/#hybrid-search-combining-sparse-and-dense-vectors
https://ar5iv.labs.arxiv.org/html/2109.10086


## Matching approach 
LLM-assisted lexical normalization and synonym expansion for domain-specific query enrichment.
using LLM to normalize and suggest synonyms for domain-specific queries.


## Ngrams
